### PR TITLE
Add support for clang and OS X to qserv scons (DM-3662)

### PR DESCRIPTION
**Pull request by andy-slac**
Compiler-dependent options were moved from state.py to compiler tool
which detects specific platforms and compilers and sets compile/link
options.

### CODE REVIEW

File: site_scons/site_tools/compiler.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
+"""swig_scanner
+
+Special tool wheich defines compiling/linking options.
```
Comment (timj) [inline site_scons/site_tools/compiler.py]: Minor typo here.

File: site_scons/site_tools/compiler.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
+"""swig_scanner
```
Comment (timj) [inline site_scons/site_tools/compiler.py]: Is it really `swig_scanner`?

File: site_scons/site_tools/compiler.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+def _compiler(env):
+    """
+    Guess compiler type and version. Returns pair of strings:
+    - toolchain name, one of 'gcc', 'clang'
+    - compiler + version, e.g. 'gcc44', 'gcc48', 'clang35'
+    """
+
+    # for qserv we only set or use CXX, ignore CC
+    comp = os.path.basename(env['CXX'])
```
Comment (timj) [inline site_scons/site_tools/compiler.py]: I don't think `comp` is used any more.

File: site_scons/site_tools/compiler.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+def generate(env):
+
+    # current platform
+    platform = sys.platform
+
+    # compiler
+    toolchain, comp_version = _compiler(env)
+
+    state.log.info("Building on %s with %s" % (platform, toolchain))
```
Comment (timj) [inline site_scons/site_tools/compiler.py]: This message will ignore the `-Q` option to `scons`. I think that's consistent with all the messages coming out of the Qserv {{scons}} but maybe in the future `-Q` could be used.

File: site_scons/site_tools/compiler.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+            # gcc44 only supports c++0x
+            env.Append(CXXFLAGS=['-std=c++0x'])
+        else:
+            # newer compilers are expected to be c++11
+            env.Append(CXXFLAGS=['-std=c++11'])
+
+        # to make shared libraries link correctly we need -rpath-link option, for now add everything
+        # that is in LD_LIBRARY_PATH
+        if 'LD_LIBRARY_PATH' in os.environ:
+            env.Append(LINKFLAGS = ["-Wl,-rpath-link="+os.environ["LD_LIBRARY_PATH"]])
```
Comment (timj) [inline site_scons/site_tools/compiler.py]: Minor nit: PEP8 doesn't like spaces between keyword and argument.

File: site_scons/site_tools/compiler.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        env.Append(LINKFLAGS=["-pthread"])
+
+    else:
+
+        raise SCons.Errors.StopError('Unsupported platform or toolchain: %s/%s' % (platform, toolchain))
+
+    # rest works for any compiler/platform
+    if env['debug'] == 'yes':
+        state.log.info("Debug build flag (-g) requested.")
+        env.Append(CCFLAGS = ['-g'])
```
Comment (timj) [inline site_scons/site_tools/compiler.py]: PEP8 ' = '

File: site_scons/site_tools/compiler.py (hunk 2 above)
Comment (andy-slac) [inline site_scons/site_tools/compiler.py]: It was at some point :) Stupid copy/paste error, will fix

File: site_scons/site_tools/compiler.py (hunk 1 above)
Comment (andy-slac) [inline site_scons/site_tools/compiler.py]: OK.

File: site_scons/site_tools/compiler.py (hunk 3 above)
Comment (andy-slac) [inline site_scons/site_tools/compiler.py]: True, will remove.

File: site_scons/site_tools/compiler.py (hunk 4 above)
Comment (andy-slac) [inline site_scons/site_tools/compiler.py]: We do need better logging there, it's in my long list of improvements (unless we go cmake route).

File: site_scons/site_tools/compiler.py (hunk 5 above)
Comment (andy-slac) [inline site_scons/site_tools/compiler.py]: OK. My auto-formatter decided to skip it for some reason.

File: site_scons/site_tools/compiler.py (hunk 6 above)
Comment (andy-slac) [inline site_scons/site_tools/compiler.py]: Thanks

**Pull request merged**